src/lib/storage/table.cpp

```diff
@@ -81,23 +83,71 @@ const std::string& Table::column_name(const ColumnID column_id) const { return _
 
 const std::string& Table::column_type(const ColumnID column_id) const { return _col_types.at(column_id); }
 
-Chunk& Table::get_chunk(ChunkID chunk_id) { return *_chunks.at(chunk_id); }
+Chunk& Table::get_chunk(ChunkID chunk_id) {
+  std::lock_guard<std::mutex> lock(_chunk_lock);
+  return *_chunks.at(chunk_id);
+}
+
+const Chunk& Table::get_chunk(ChunkID chunk_id) const {
+  std::lock_guard<std::mutex> lock(_chunk_lock);
+  return std::as_const(_get_chunk(chunk_id));
+}
 
-const Chunk& Table::get_chunk(ChunkID chunk_id) const { return *_chunks.at(chunk_id); }
+Chunk& Table::_get_chunk(ChunkID chunk_id) const { return *_chunks.at(chunk_id); }
 
 void Table::print(std::ostream& out) const {
   int col_width = 20;
-  for (auto name : _col_names) {
+  for (const auto& name : _col_names) {
     std::string value = name;
     if (static_cast<int>(value.length()) > col_width) value.resize(col_width);
     out << value << std::string(col_width - value.length(), ' ');
   }
   out << "\n" << std::string(col_width * column_count(), '-') << "\n";
-  for (auto chunk : _chunks) {
+  for (const auto& chunk : _chunks) {
     chunk->print(col_width);
   }
 }
 
-void Table::compress_chunk(ChunkID chunk_id) { throw std::runtime_error("Implement Table::compress_chunk"); }
+void Table::compress_chunk(ChunkID chunk_id) {
+  std::lock_guard<std::mutex> lock(_chunk_lock);
+  auto& uncompressed_chunk = _get_chunk(chunk_id);
+  Assert(uncompressed_chunk.size() == target_chunk_size(),
+         "Attempt to compress chunk that is not yet completely filled.");
+
+  std::shared_ptr<Chunk> compressed_chunk = _compress_multithreaded(uncompressed_chunk);
+  _chunks[chunk_id] = compressed_chunk;
+}
+
+std::shared_ptr<Chunk> Table::_compress_multithreaded(Chunk& uncompressed_chunk) {
+  auto compressed_chunk = std::make_shared<Chunk>();
+  auto col_count = column_count();
+  std::vector<std::thread> column_threads = {};
+  column_threads.reserve(col_count);
+
+  for (ColumnID column_id = ColumnID{0}; column_id < col_count; column_id++) {
+    std::thread single_column_thread(&Table::_compress_column, this, std::ref(uncompressed_chunk),
+                                     std::ref(compressed_chunk), column_id);
+    column_threads.push_back(std::move(single_column_thread));
+  }
+
+  for (std::thread& column_thread : column_threads) {
+    if (column_thread.joinable()) {
+      column_thread.join();
+    }
+  }
+
+  return compressed_chunk;
+}
+
+void Table::_compress_column(Chunk& uncompressed_chunk, std::shared_ptr<Chunk>& compressed_chunk,
+                             ColumnID col_id) const {
+  const auto& column_segment = uncompressed_chunk.get_segment(col_id);
+
+  resolve_data_type(column_type(col_id), [&](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
+    const auto compressed_segment = std::make_shared<DictionarySegment<ColumnDataType>>(column_segment);
+    compressed_chunk->add_segment(compressed_segment);
```

Comment:
`add_segment` could theoretically be called from multiple concurrent threads at the same time. See Moodle for a [similar](https://moodle.hpi.de/mod/hsuforum/discuss.php?d=43) discussion.